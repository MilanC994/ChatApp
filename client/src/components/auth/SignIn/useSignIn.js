import { useState, useEffect, useCallback } from 'react'
import SignUserIn from '../../mutations/SignUserIn'
import { emailValidation, passwordValidation } from '../../../utils/credentialsValidation'
import useInputField from '../../../hooks/useInputField'
import validateForm from '../../../utils/validateForm'


const useSignIn = ({ inviteId, inviteEmail }) => {
    useEffect(() => {
        if (localStorage.getItem('token')) {
        setSignedIn(true)
        }
    },[])


    const [signedIn, setSignedIn] = useState(false)
    const [error, setError] = useState(false)

    const emailValue = inviteEmail ? inviteEmail : ''

    const { input: emailField, data: email } = useInputField('email', emailValue, emailValidation)
    const { input: passwordField, data: password} = useInputField('password','', passwordValidation)

    const handleSubmit = useCallback(async () => {
        console.log("HANDLE SUBMIT")
        if(!validateForm([email, password])){ console.log("validation res   je false"); return}
        SignUserIn(
            emailField.value,
            passwordField.value,
            inviteId,
            async (token, error) => {
                if (error) {
                    setError(error)
                } else {
                    await localStorage.setItem('token', token)
                    setError(null)
                    setSignedIn(true)

                }
            }
        )
    },[emailField, passwordField, inviteId])

    return { emailField, passwordField, handleSubmit, error, signedIn }
}

export default useSignIn