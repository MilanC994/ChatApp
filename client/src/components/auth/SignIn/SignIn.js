import React, { useState, useEffect } from 'react'
import Avatar from '@material-ui/core/Avatar'
import Button from '@material-ui/core/Button'
import CssBaseline from '@material-ui/core/CssBaseline'
import TextField from '@material-ui/core/TextField'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import Checkbox from '@material-ui/core/Checkbox'
import Link from '@material-ui/core/Link'
import Grid from '@material-ui/core/Grid'
import LockOutlinedIcon from '@material-ui/icons/LockOutlined'
import { Alert } from '@material-ui/lab'
import Typography from '@material-ui/core/Typography'
import { withStyles } from '@material-ui/core/styles'
import Container from '@material-ui/core/Container'
import { Redirect } from 'react-router-dom'
import Navbar from '../../nav/Navbar/Navbar'
import SignInStyle from './SignInStyle'
import useSignIn from './useSignIn'


const SignIn = ({ classes, inviteId, inviteEmail }) => {

  const { emailField: email, passwordField: password, handleSubmit, error, signedIn } = useSignIn({ inviteId, inviteEmail })
 
    if (signedIn) {
      return <Redirect to={'/dashboard'} />
    }
    return (
      <React.Fragment>
        {!inviteId && <Navbar />}
        <Container component='main' maxWidth='xs'>
          <CssBaseline />
          <div className={classes.paper}>
            <Avatar className={classes.avatar}>
              <LockOutlinedIcon />
            </Avatar>
            <Typography component='h1' variant='h5'>
              Sign in
            </Typography>
            <form className={classes.form} noValidate>
              <TextField
                variant='outlined'
                margin='normal'
                required
                fullWidth
                id='email'
                label='Email Address'
                autoComplete='email'
                autoFocus
                {...email}
                error={!!email.error}
                disabled={!!inviteEmail}
              />
              {!!email.error &&<Alert severity="error">{email.error}</Alert>}
              <TextField
                variant='outlined'
                margin='normal'
                required
                fullWidth
                label='Password'
                type='password'
                id='password'
                {...password}
                error={!!password.error}
              />
              {!!password.error && <Alert severity="error">{password.error}</Alert>}
              <FormControlLabel
                control={<Checkbox value='remember' color='primary' />}
                label='Remember me'
              />
              <Button
                fullWidth
                variant='contained'
                color='primary'
                className={classes.submit}
                onClick={handleSubmit}
                id='signInBtn'
              >
                Sign In
              </Button>
              <Grid container>
                <Grid item xs>
                  <Link href='#' variant='body2'>
                    Forgot password?
                  </Link>
                </Grid>
                <Grid item>
                  <Link href='#' variant='body2'>
                    {"Don't have an account? Sign Up"}
                  </Link>
                </Grid>
              </Grid>
            </form>
          </div>
          {error && <Alert severity="error">{error}</Alert>}
        </Container>
      </React.Fragment>
    )
}
export default withStyles(SignInStyle)(SignIn)
