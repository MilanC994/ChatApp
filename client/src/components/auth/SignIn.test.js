import React from 'react'
import { shallow } from 'enzyme'
import SignIn from './SignIn'
import toJson from 'enzyme-to-json'
import cridentialsValidation from '../utils/cridentialsValidation'

describe('SignIn Component test', () => {
  const wrapper = shallow(<SignIn />)

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  //Mock funkcije
  it('should have an empty email and password state', () => {
    //Optionally test to check if password and email are empty strings on
    //  setup
    expect(wrapper.dive().state('email')).toEqual('')
    expect(wrapper.dive().state('password')).toEqual('')
  })

  it('test email and password pressence', () => {
    //no input
    expect(cridentialsValidation('name', '', '', 'login')).toEqual(false)
    //wrong email
    expect(cridentialsValidation('name', 'wrongEmail', '', 'login')).toEqual(
      false
    )

    expect(
      cridentialsValidation('name', 'email@email.com', 'somepassword', 'login')
    ).toEqual(true)
  })

  it('onButtonClick sign user in or output error', () => {
    wrapper.dive().setState({
      email: 'goodemail@email.com',
      password: 'somegoodpassword',
    })

    wrapper.dive().find('#signInBtn').simulate('click')
    expect(wrapper.dive().state('signedIn')).toEqual(true)
  })
})
