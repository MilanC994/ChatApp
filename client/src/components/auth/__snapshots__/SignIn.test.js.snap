// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn Component test renders correctly 1`] = `
<SignIn
  classes={
    Object {
      "avatar": "SignIn-avatar-2",
      "form": "SignIn-form-3",
      "paper": "SignIn-paper-1",
    }
  }
/>
`;
